### ISSUE TITLE
method-unbinding errors prevent safe browser compatibility checks

### ISSUE THREAD

**Issue by jrkatz** [facebook/flow#8705]
Flow version: v0.155.0

## Expected behavior
We should be able to check that browser APIs are available without running afoul of method-unbinding errors. For example, `if (typeof navigator?.permissions?.query === 'function') { ... ; navigator.permissions.query(...); ... }` should not produce a flow error.

## Actual behavior
The above produces a method-unbinding error at the reference to `query`: 
```
property query [1] cannot be unbound from the context [2] where it was defined. [method-unbinding]
```

Only lenient checks of the form `if (navigator && navigator.permissions && 'query' in navigator.permissions)` appear possible, but these checks are wordy and incomplete: they will not handle the possibility the field in question exists but is not a function, and typos in the property name will not be caught.

It's one thing to suppress issues here and there, but this is a very common pattern and we would like to avoid having our codebase littered with `// $FlowFixMe` comments. That encourages devs to treat _real_ errors caught by flow as _flow_ errors and move on.

If this is unfixable for some reason, I am also eager to hear any clever alternatives I have failed to discover.

* Link to Try-Flow or Github repo: https://flow.org/try/#0PTAqCcFcDsGMEMAuBTAJgAgHLwG4EsBzJAe3HQDJ0AFZcAWzwGdG9jpGAadAGzwDMUGPuGJ10AC0SIADowBcIAnkTjIAIwB0sUcD7xYyNcWIBrXd2IB3YGotrgdeIxThgve0boaAVowBQqMiw3PDgyOjBTozUtAzMrOzoAN7oqWnpGZmpfqkAjpC0AJ4AFFll5RWVmTlp0rFMLGwAIsiMsHLoDAAeaDUAlB1UInHIADzdaAB8ANx+AL4BQSFhESHMWLiEJGRJNal19A0J8jGH8WyM0+nzi5ErOKHo0JtEiKQd2PivpLN+2uyIdChAjoAC8ySe8DoyA6AHICMhiBYEIgErD0HNfiB0KBxEx0GF8ngwtF4OhGJBpNISY1oOg+KROsgVMRUABaGBqPDQVDcgh+fjoYqIQp1Yh8SFfbYAfg0BzitMYsvyRTBoPBsL4MFgqLYsL6yRq2IAypTpMRGOFLOFxLhwm90KhiJ1SPbbXTvJBnBF4NxuOgVOFoSzUBwjcByc7iNJdc9-bBbdzorwTOFYc8pW9wLL5Uc2MqCuBCrLihoy8DGAbQuF4OGnmw2dHYzV-owkcgNBYCMVNcZ9Vcahmtlm5fVzuwNCqi8VgX0NIHoMVigbQZNknM+rMFtjceEAAbcvfoVOFSykDBn8AmTjoNSQQF6PDcaLcn2W6IqJAB8ThB7cAroPi0DEICZJanAsaAXSbbQreIiWJaZA4LQioaAKEo9lOhToq+Q7fOAo5nIqBq7Gkrbtp2xDdr2xD9ns2KCuIoogT+qIIH6hToLCWE4dEwGgfS2qQXegIweELG0AGorhK+YlAnAeDINAgJqPBiFcNasIrK+iAiOo3DhMBlhoWkeHbIRCrHJOhYlLO84-ouy5gmuSQbluWIRru6AHtAR6+m26DUsgyHKdEfAWJY9LDD6iAJnyUnmvI6FCtxyCQEWOF0mZI65uOlYQti1roGekDcBgKZuviDqPtwezkQZlHUQytGbns2WkBZeYTlhM7gAQc4LkuK4uW5Nx+NiADCbDIeAlrcIUXAqPiKIJq0372sxXAieg4oRMQ6WWoBEpFXweBdIBgKaX6Eh2kCTIhhy0Bcjy8W0CIZDwEQ3LJcK0m7e12adXlyppUWaoauBOpoiRdUXBRXY9s1dGmS85m5ahPV2YNTmruurVzEAA

**Comment by charlag** [comment]
It was discussed in #8689, they could probably be merged

**Comment by jrkatz** [comment]
> It was discussed in #8689, they could probably be merged

Agreed, conversation on that one trended in this direction, though the originally cited issue is different. Not sure how the flow team feels about consolidating in that way. They've also reopened some issues very similar to this one that were closed/wontfixed when I raised this issue. (#8686)

Optimistically, all these issues will soon be consolidated to a single issue: changing the `method-unbinding` error to a `method-rebinding` error; i.e. only reporting an error if an unbound method is rebound and/or called, and reporting that error at the point of use instead of at the point of unbinding. I'm sure that's much more complicated an implementation than the current method-unbinding implementation, but I believe it addresses the unsoundness issue while allowing more innocuous patterns like checking type and existence, and being more friendly to the kind of decoration & wrapping that testing libraries do.

**Comment by dsainati1** [comment]
This is on our radar but at the moment it isn't a priority. We are working on a large restructuring of Flow's type inference, and one of the side effects of this work is that it will make special-case behavior like this much easier, but at the moment it is challenging to make ad-hoc adjustments like this. This will get fixed eventually, but it is unlikely that I will get to it soon.

**Comment by jrkatz** [comment]
@dsainati1 Thanks for the update. I'm happy so long as it will probably be resolved one day and so long as I have this issue to point to when people as me why things are the way they are. Have fun with the big restructuring.

**Comment by SamChou19815** [comment]
no longer errors
